proposals/lotus-blockstore-interface.md

```diff
@@ -0,0 +1,44 @@
+# Extended Blockstore Interface
+
+Authors: vyzo
+
+## What is the problem this project solves?
+
+We currently have a limited "dumb" blockstore interface, which does not allow us to pass context options so as to inform the blockstore about certain properties of objects being stored.
+
+This is particularly important in the context of the splitstore, where we have the following problems:
+- We lack the ability to inform the splitstore of the actual epoch an object belongs, which requires us to "guess" by estimating based int he current tipset and wall clock (see [#6474](https://github.com/filecoin-project/lotus/pull/6474)). This is problematic for many reasons, as an object might be misassigned its epoch (and its completely broken during catch up sync in current master).
+- We lack the ability to inform the splitstore when an access is initiated from the network. In conjunction with the transactional gc logic in [#6474](https://github.com/filecoin-project/lotus/pull/6474) this would result in network requests keeping objects that would otherwise be pruged from the hotstore live.
+This is both a performance/gc issue (the hotstore is larger than it needs to be) but it is also a potential attack vector for nodes that need to run with fixed hardware (e.g. network boosters).
+- We lack the ability to inform the splitstore that certain objects should be reified from cold storage to hot storage. For instance, a miner doing tipset state computation in an older tipset may want to keep this in the hot storage. It also means that misses result in permanent banishment in the coldstore and makes it hard to gc by simply removing the coldstore.
+
+## Impact
+
+This will make the splitstore a much more reliable component, working towards making it the default blockstore in lotus.
+
+## The idea
+
+Extend the blockstore interface so as to accept an `Options` struct, which specifies the necessary context constraints.
+
+Methods would look like this:
+```
+type Blockstore interface {
+  ...
+  Get(c cid.Cid, opts Options) (blocks.Block, error)
```

Comment:
Why `opts Options` instead of `opts Options...`? Are there only options for Get, what about the other operations on the blockstore?

Reply:
so that we don't have to parse basically.

Reply:
Also, it's for all methods, not just Get -- I just didnt want to list them all here.

And and and, I really want to kill the `AllKeysChan` atrocity and replace it with `ForEach`.

Reply:
This is a low-level operation, we can't afford to incur in the cost of heap-escaping functional options for every invocation.

Reply:
I think this proposal would be a lot stronger with a concrete list of options that would apply to a Get operation. Currently it appears to be a catch-all for any possible future behavior.

I'm also moderately perturbed by having per-operation options rather than per-store. This implies a potential large variation in performance characteristics per operation.